File: app/src/main/java/com/genonbeta/TrebleShot/app/GroupEditableListFragment.java
package com.genonbeta.TrebleShot.app;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.collection.ArrayMap;

import com.genonbeta.TrebleShot.R;
import com.genonbeta.TrebleShot.adapter.ImageListAdapter;
import com.genonbeta.TrebleShot.widget.GroupEditableListAdapter;

import java.util.Map;

/**
 * created by: veli
 * date: 30.03.2018 16:10
 */

public abstract class GroupEditableListFragment<T extends GroupEditableListAdapter.GroupEditable, V extends GroupEditableListAdapter.GroupViewHolder, E extends GroupEditableListAdapter<T, V>>
        extends EditableListFragment<T, V, E>
{
    private Map<String, Integer> mGroupingOptions = new ArrayMap<>();
    private int mDefaultGroupingCriteria = GroupEditableListAdapter.MODE_GROUP_BY_NOTHING;

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState)
    {
        super.onViewCreated(view, savedInstanceState);
        getAdapter().setGroupBy(getGroupingCriteria());
    }

    @Override
    public int onGridSpanSize(int viewType, int currentSpanSize)
    {
        return viewType == ImageListAdapter.VIEW_TYPE_REPRESENTATIVE
                || viewType == ImageListAdapter.VIEW_TYPE_ACTION_BUTTON
                ? currentSpanSize
                : super.onGridSpanSize(viewType, currentSpanSize);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater)
    {
        super.onCreateOptionsMenu(menu, inflater);

        Map<String, Integer> options = new ArrayMap<>();

        onGroupingOptions(options);

        mGroupingOptions.clear();
        mGroupingOptions.putAll(options);

        if (mGroupingOptions.size() > 0) {
            inflater.inflate(R.menu.actions_abs_group_shareable_list, menu);
            MenuItem groupingItem = menu.findItem(R.id.actions_abs_group_shareable_grouping);

            if (groupingItem != null)
                applyDynamicMenuItems(groupingItem, R.id.actions_abs_group_shareable_group_grouping, mGroupingOptions);
        }
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu)
    {
        super.onPrepareOptionsMenu(menu);
        checkPreferredDynamicItem(menu.findItem(R.id.actions_abs_group_shareable_grouping), getGroupingCriteria(), mGroupingOptions);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        if (item.getGroupId() == R.id.actions_abs_group_shareable_group_grouping)
            changeGroupingCriteria(item.getOrder());
        else
            return super.onOptionsItemSelected(item);

        return true;
    }

    public void onGroupingOptions(Map<String, Integer> options)
    {
    }

    public void changeGroupingCriteria(int criteria)
    {
        getViewPreferences().edit()
                .putInt(getUniqueSettingKey("GroupBy"), criteria)
                .apply();

        getAdapter().setGroupBy(criteria);

        refreshList();
    }

    public int getGroupingCriteria()
    {
        return getViewPreferences()
                .getInt(getUniqueSettingKey("GroupBy"), mDefaultGroupingCriteria);
    }

    public void setDefaultGroupingCriteria(int groupingCriteria)
    {
        mDefaultGroupingCriteria = groupingCriteria;
    }
}

File: app/src/main/java/com/genonbeta/TrebleShot/app/GroupEditableListFragment.kt
/*
 * Copyright (C) 2019 Veli Tasalı
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.genonbeta.TrebleShot.app

import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import androidx.collection.ArrayMap
import com.genonbeta.TrebleShot.R
import com.genonbeta.TrebleShot.widget.GroupEditableListAdapter
import com.genonbeta.TrebleShot.widget.GroupEditableListAdapter.GroupEditable
import com.genonbeta.TrebleShot.widget.GroupEditableListAdapter.GroupViewHolder

/**
 * created by: veli
 * date: 30.03.2018 16:10
 */
abstract class GroupEditableListFragment<T : GroupEditable, V : GroupViewHolder, E : GroupEditableListAdapter<T, V>>() :
    EditableListFragment<T, V, E>() {
    override var adapter: E
        get() = super.adapter
        set(value) {
            super.adapter = value
            value.setGroupBy(groupingCriteria)
        }

    private val groupingOptions: MutableMap<String, Int> = ArrayMap()

    var defaultGroupingCriteria: Int = GroupEditableListAdapter.MODE_GROUP_BY_NOTHING

    override fun onGridSpanSize(viewType: Int, currentSpanSize: Int): Int {
        return if (viewType == GroupEditableListAdapter.VIEW_TYPE_REPRESENTATIVE
            || viewType == GroupEditableListAdapter.VIEW_TYPE_ACTION_BUTTON
        ) currentSpanSize else super.onGridSpanSize(viewType, currentSpanSize)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        if (!isUsingLocalSelection || !isLocalSelectionActivated) {
            val options: MutableMap<String, Int> = ArrayMap()
            onGroupingOptions(options)
            groupingOptions.clear()
            groupingOptions.putAll(options)
            if (groupingOptions.isNotEmpty()) {
                inflater.inflate(R.menu.actions_abs_group_shareable_list, menu)
                applyDynamicMenuItems(
                    menu.findItem(R.id.actions_abs_group_shareable_grouping),
                    R.id.actions_abs_group_shareable_group_grouping,
                    groupingOptions
                )
            }
        }
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
        if (!isUsingLocalSelection || !isLocalSelectionActivated) {
            checkPreferredDynamicItem(
                menu.findItem(R.id.actions_abs_group_shareable_grouping), groupingCriteria,
                groupingOptions
            )
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.groupId == R.id.actions_abs_group_shareable_group_grouping)
            groupingCriteria = item.order
        else
            return super.onOptionsItemSelected(item)
        return true
    }

    open fun onGroupingOptions(options: MutableMap<String, Int>) {}

    var groupingCriteria: Int
        get() = viewPreferences.getInt(getUniqueSettingKey("GroupBy"), defaultGroupingCriteria)
        set(value) {
            viewPreferences.edit()
                .putInt(getUniqueSettingKey("GroupBy"), value)
                .apply()
            adapter.setGroupBy(value)
            refreshList()
        }
}